rfcs/2021-09-21-9292-throttle-transform.md

```diff
@@ -0,0 +1,149 @@
+per second as using facing configuration (or expand to more time frames) and translate that into a `governor::Quota` for internal use. We should not expose
+the `Quota` directly to users to ensure ease of configuration for end users.
+
+Config:
+
+```rust
+pub struct ThrottleConfig {
+    events_per_second: u32,
+    bytes_per_second: u32,
+    key_field: Option<String>,
+    exclude: Option<AnyCondition>,
+}
+```
+
+TaskTransform:
+
+```rust
+impl TaskTransform for Throttle {
+    fn transform(
+        self: Box<Self>,
+        mut input_rx: Pin<Box<dyn Stream<Item = Event> + Send>>,
+    ) -> Pin<Box<dyn Stream<Item = Event> + Send>>
+    where
+        Self: 'static,
+    {
+        let lim = RateLimiter::keyed(Quota::per_second(self.events_per_second));
+
+        let mut flush_stream = tokio::time::interval(Duration::from_millis(1000));
+
+        Box::pin(
+            stream! {
+              loop {
+                let mut output = Vec::new();
+                let done = tokio::select! {
+                    _ = flush_stream.tick() => {
+                        false
+                    }
+                    maybe_event = input_rx.next() => {
+                        match maybe_event {
+                            None => true,
+                            Some(event) => {
+			    	if let Some(condition) = self.exclude_as_ref() {
+				    if condition.check(&event) {
+				        output.push(event);
+					false
+				    }
+				}
+
+				let value = self
+				    .key_field
+				    .as_ref()
+				    .and_then(|key_field| event.get(key_field))
+				    .map(|v| v.to_string_lossy());
+
+                                match lim.check_key_n(value, size) {
+                                    Ok(()) => {
+                                        output.push(event);
+                                        false
+                                    }
+                                    _ => {
+                                        // Dropping event
+                                        false
+                                    }
+                                }
+                            }
+                        }
+                    }
+                };
+                yield stream::iter(output.into_iter());
+                if done { break }
+              }
+            }
+            .flatten(),
+        )
+    }
+}
+```
+
+## Rationale
+
+- Controlling throughput provides better reliability for downstream services which can be critical in observability platforms
+- Setting granular quotas is invaluable as an administrator running Vector for a number of users to ensure fair use of the pipeline
+
+## Drawbacks
+
+- When rate limiting by `Bytes` in a transform it will be the unserialized form of the event, which will differ from what downstream sinks will actually receive
```

Comment:
I think it's worth mentioning, but I wouldn't call this a drawback. It's inevitable, given that each sink choses how to encode the events (e.g. whether to apply compression), so transforms will always have to work with the unserialized event size.

Reply:
Compression is a good point and likely more impactful than serialization - it'd be quite likely that if we `throttle`d as a transform and compressed payloads to the `sink` we'd run into many situations where we're not going to saturate that service and underperform.